encoding/ccf/decode_type.go

```diff
@@ -0,0 +1,615 @@
+	borrowType, err := decodeTypeFn(types)
+	if err != nil {
+		return nil, err
+	}
+
+	return cadence.NewMeteredCapabilityType(d.gauge, borrowType), nil
+}
+
+// decodeReferenceType decodes reference-type or reference-type-value as
+// language=CDDL
+// reference-type =
+//
+//	; cbor-tag-reference-type
+//	#6.142([
+//	  authorized: bool,
+//	  type: inline-type,
+//	])
+//
+// reference-type-value =
+//
+//	; cbor-tag-reference-type-value
+//	#6.190([
+//	  authorized: bool,
+//	  type: type-value,
+//	])
+//
+// NOTE: decodeTypeFn is responsible for decoding inline-type or type-value.
+func (d *Decoder) decodeReferenceType(
+	types cadenceTypeByCCFTypeID,
+	decodeTypeFn decodeTypeFn,
+) (cadence.Type, error) {
+	// Decode array head of length 2
+	err := decodeCBORArrayWithKnownSize(d.dec, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// element 0: authorized
+	authorized, err := d.dec.DecodeBool()
+	if err != nil {
+		return nil, err
+	}
+
+	// element 0: type (inline-type or type-value)
+	elementType, err := decodeTypeFn(types)
+	if err != nil {
+		return nil, err
+	}
+
+	return cadence.NewMeteredReferenceType(d.gauge, authorized, elementType), nil
+}
+
+// decodeRestrictedType decodes restricted-type or restricted-type-value as
+// language=CDDL
+// restricted-type =
+//
+//	; cbor-tag-restricted-type
+//	#6.143([
+//	  cadence-type-id: cadence-type-id,
+//	  type: inline-type,
+//	  restrictions: [+ inline-type]
+//	])
+//
+// restricted-type-value =
+//
+//	; cbor-tag-restricted-type-value
+//	#6.191([
+//	  cadence-type-id: cadence-type-id,
+//	  type: type-value,
+//	  restrictions: [+ type-value]
+//	])
+//
+// NOTE: decodeTypeFn is responsible for decoding inline-type or type-value.
+func (d *Decoder) decodeRestrictedType(
+	types cadenceTypeByCCFTypeID,
+	decodeTypeFn decodeTypeFn,
+) (cadence.Type, error) {
+	// Decode array of length 3.
+	err := decodeCBORArrayWithKnownSize(d.dec, 3)
+	if err != nil {
+		return nil, err
+	}
+
+	// element 0: cadence-type-id
+	typeID, err := d.dec.DecodeString()
+	if err != nil {
+		return nil, err
+	}
+
+	// element 1: type
+	typ, err := decodeTypeFn(types)
+	if err != nil {
+		return nil, err
+	}
+
+	// element 2: restrictions
+	restrictionCount, err := d.dec.DecodeArrayHead()
+	if err != nil {
+		return nil, err
+	}
+
+	restrictionTypeIDs := make(map[string]struct{}, restrictionCount)
+	var previousRestrictedTypeID string
+
+	restrictions := make([]cadence.Type, restrictionCount)
+	for i := 0; i < int(restrictionCount); i++ {
+		// Decode restriction.
+		restrictedType, err := decodeTypeFn(types)
+		if err != nil {
+			return nil, err
+		}
+
+		// "Valid CCF Encoding Requirements" in CCF specs:
+		//
+		//   "Elements MUST be unique in restricted-type or restricted-type-value."
+		if _, ok := restrictionTypeIDs[restrictedType.ID()]; ok {
+			return nil, fmt.Errorf("found duplicate restricted type %s", restrictedType.ID())
+		}
+
+		// "Deterministic CCF Encoding Requirements" in CCF specs:
```

Comment:
For a follow-up PR, in general: Maybe we can make the deterministic encoding checking in the decoder optional? Most of the time we do not need, i.e. we could avoid the associated overhead

Reply:
Yeah, I agree.  I previously looked into this and there were some complications.  We can tackle it in a follow-up PR as suggested.